proposals/lip-0045.md

```diff
@@ -984,164 +834,351 @@ The Interoperability module contains the following commands:
+            )
+            raise Exception('Receiving chain does not exist.')
+        partnerChainID = receivingChainID
+    # Processing on a sidechain
+    else:
+        # Check for direct channel
+        if chainAccount(receivingChainID) does not exists:
+            partnerChainID = mainchainID
+        else:
+            partnerChainID = receivingChainID
 
-##### Returns
+    # Chain must be live; This checks is always on the receivingChainID
+    if not isLive(receivingChainID):
+        emitPersistentEvent(
+            module = MODULE_NAME_INTEROPERABILITY,
+            name = EVENT_NAME_CCM_SENT,
+            data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_CHANNEL_UNAVAILABLE},
+            topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
+        )
+        raise Exception('Receiving chain is not live.')
+    # Chain status must be active
+    if not chainAccount(partnerChainID).status == CHAIN_STATUS_ACTIVE:
+        emitPersistentEvent(
+            module = MODULE_NAME_INTEROPERABILITY,
+            name = EVENT_NAME_CCM_SENT,
+            data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_CHANNEL_UNAVAILABLE},
+            topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
+        )
+        raise Exception('Receiving chain is not active.')
 
-This function returns `channel(chainID)`.
+    # Create cross-chain message
+    paramsSchema = JSON schema associated with (module, crossChainCommand)
+    ccm = {
+        "nonce": ownChainAccount.nonce,
+        "module": module,
+        "crossChainCommand": crossChainCommand,
+        "sendingChainID": ownChainAccount.chainID,
+        "receivingChainID": receivingChainID,
+        "fee": fee,
+        "status": status,
+        "params": encode(paramsSchema, params)
+    }
 
+    # Validate ccm size
+    try:
+        validateFormat(ccm)
+    except Exception as e:
+        emitPersistentEvent(
+            module = MODULE_NAME_INTEROPERABILITY,
+            name = EVENT_NAME_CCM_SENT,
+            data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_INVALID_FORMAT},
+            topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
+        )
+        raise e
+
+    # Pay message fee
+    if fee > 0:
+        try:
+            Token.payMessageFee(sendingAddress, fee)
+        except Exception as e:
+            emitPersistentEvent(
+                module = MODULE_NAME_INTEROPERABILITY,
+                name = EVENT_NAME_CCM_SENT,
+                data = {"ccmID": ccmID, "status": CCM_SENT_STATUS_MESSAGE_FEE_ERROR},
+                topics = [ccm.sendingChainID, ccm.receivingChainID, ccmID]
+            )
+            raise e
 
-#### getOwnChainAccount
+    addToOutbox(partnerChainID, ccm)
+    ownChainAccount.nonce += 1
 
-The `getOwnChainAccount` function returns the object stored in the own chain data substore.
+    # Emit CCM processed event.
+    ccmID = sha256(encode(crossChainMessageSchema, ccm))
```

Comment:
You only compute the CCM ID here, but use it in all events before. Also it may not make sense/be possible to compute the ID if the format validation fails, so maybe the CCM ID needs to be empty there. If you want to use the CCM ID in events, consider doing the format validation in the send function first.

Reply:
yes, there were several small imprecisions here. Please take a look now.